Rewrite assert statement with torch._assert under config

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88246

This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
Rewrite assert statement with torch._assert under config

[ghstack-poisoned]

jansel commented on test/dynamo/test_repros.py:
```diff
@@ -1758,6 +1758,36 @@ def forward(self, inp):
+    def test_rewrite_assert(self):
+        def f(x):
+            b = x.sin()
+            assert x[0] == 3
```
Can you add a few more test cases:
```
# should get constant propped away (no assert in the graph)
assert x.dtype == torch.float32
assert True
# has message
assert assert x[0] == 3, "foo"
# has message with format
assert assert x[0] == 3, f"foo {x[0]}"
```

Don't need to handle the second two, but we should run correctly on them.

tugsbayasgalan replied:
will do

tugsbayasgalan replied:
done

jansel commented on torch/_dynamo/config.py:
```diff
@@ -31,7 +31,7 @@
 
 # Verbose will print full stack traces on warnings and errors
-verbose = False
+verbose = True
```
!

tugsbayasgalan replied:
oops my bad

jansel commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
     def inner(self: "InstructionTranslatorBase", inst: Instruction):
         value: VariableTracker = self.pop()
         self.output.guards.update(value.guards)
+        if config.rewrite_assert_with_torch_assert:
```
Pull this into a helper method.

tugsbayasgalan replied:
done

tugsbayasgalan replied:
done

jansel commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
+                current_instruction_pointer = self.instruction_pointer
+                inst = self.instructions[current_instruction_pointer]
+                # Detect LOAD_ASSERTION_ERROR
+                if inst.opname == "LOAD_ASSERTION_ERROR":
```
This bytecode was added in 3.9, so might not work in 3.8/3.7

suo replied:
yeah, in particular fbcode uses 3.8, where it looks unfortunately more generic (LOAD_GLOBAL of AssertionError). So I would expect to see a Python version check and different handling here.

tugsbayasgalan replied:
done

tugsbayasgalan replied:
done

jansel commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
+                    if current_instruction_pointer < len(self.instructions):
+                        inst_for_raise = self.instructions[current_instruction_pointer]
+                        # DETECT RAISE_VARARGS
+                        if inst_for_raise.opname == "RAISE_VARARGS":
```
If this were a helper function, you could return rather than deeply nesting if statements.

tugsbayasgalan replied:
done

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
+                current_instruction_pointer = self.instruction_pointer
+                inst = self.instructions[current_instruction_pointer]
+                # Detect LOAD_ASSERTION_ERROR
+                if inst.opname == "LOAD_ASSERTION_ERROR":
```
hmm why is this nested in generic jump instead of a top level

`def LOAD_ASSERTION_ERROR(self, inst)` ?

tugsbayasgalan replied:
Because whenever we see JUMP_INSTRUCTION, we want to look ahead by 2 instructions to see these three form as assertion. So it is little annoying to handle it per instruction.

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
+                        inst_for_raise = self.instructions[current_instruction_pointer]
+                        # DETECT RAISE_VARARGS
+                        if inst_for_raise.opname == "RAISE_VARARGS":
+                            assert_error_msg = ConstantVariable(value="assertion failure")
```
This is not sound - the "rhs" logic of the comma for an assert can have anything in it - including logic and function calls. You need to either:

1) Trace it
2) Disallow asserts with anything except strings in that field

```
def f(x): 
    try:
        print("X is", x)
        assert(x < 3), f(1)
        return x
    except:
        pass

f(3)
```

tugsbayasgalan replied:
Sounds good. I think string is good enough for us now.

voznesenskym reviewed (changes requested):
Please add more test and more strictness to the right hand side logic

tugsbayasgalan committed:
Update on "[WIP] Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan commented:
Addressed all the comments :) @jansel @suo @voznesenskym

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
     return impl
 
 
+def _is_assert_statement(
```
Name looks like a naive check, actually has side effects (LOAD_CONST, push, etc). Rename please.

tugsbayasgalan replied:
Sounds good. Does `_detect_and_normalize_assert_statement` sound good?

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
+
+    current_instruction_pointer += 1
+
+    if current_instruction_pointer >= len(self.instructions):
+        return False
```
How can this be?

tugsbayasgalan replied:
I am just returning false because there is no way this sequence is going to be assertion as LOAD_ASSERTION_ERROR or LOAD_GLBAL be the last instruction. If we return false here, torchdynamo will switch to default execution path to handle it.

tugsbayasgalan replied:
This is just my attempt of being careful lol

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
+            return False
+        inst = self.instructions[current_instruction_pointer]
+
+    if inst.opname != "RAISE_VARARGS":
+        return False
```
Does this mean we are in LOAD_ASSERTION_ERROR but not LOAD_CONST or RAISE_VARARGS? How is that possible? Let's assert if that's the case?

tugsbayasgalan replied:
yeah it can be assert.

tugsbayasgalan replied:
i guess there is no weird case where you do LOAD_ASSERTION_ERROR and don't RAISE_VARARGS.

tugsbayasgalan replied:
I will just return false because this helper function probably shouldn't error out. If I return false, it will switch back to original dynamo code path, hence no regression.

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
+    has_error_msg = False
+    # DETECT RAISE_VARARGS or LOAD CONST
+    if inst.opname == "LOAD_CONST":
+        assert isinstance(
+            inst.argval, str
+        ), f"Assert rhs must be str to be rewritten, but found {type(inst.argval)}"
```
This will make all asserts that are like that failures, instead of just graph breaking as they do today, right? This feels like it would regress dynamo logic.

voznesenskym reviewed (changes requested):
Not sure we handle the asserts correctly here yet

tugsbayasgalan commented:
Can you elaborate on what else feels wrong besides the comments you left? Or is it good once I handle your comments?

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

jansel commented on torch/_dynamo/config.py:
```diff
@@ -87,6 +87,9 @@
 replay_record_enabled = False
 
+# Rewrite assert statement in python with torch._assert
+rewrite_assert_with_torch_assert = False
```
I'd be ok turning this on by default.  It would also be good to turn it on and run CI to see if that shakes out any more issues.

tugsbayasgalan replied:
Ok let me flip it and run the CI

tugsbayasgalan replied:
Just did

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

voznesenskym commented:
> Can you elaborate on what else feels wrong besides the comments you left? Or is it good once I handle your comments?

Just the comments :)

voznesenskym commented:
Can you run benchmarks with this diff? 

`time python benchmarks/dynamo/torchbench.py  --accuracy --backend inductor --train`

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

RUN_TORCHBENCH:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

RUN_TORCHBENCH:

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx


[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire


[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Unrelated test failure"

tugsbayasgalan committed:
Update on "Rewrite assert statement with torch._assert under config"


This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire


[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

DanilBaibak commented:
More details about the problem can be found here - [D41332298](https://www.internalfb.com/diff/D41332298)

malfet commented:
@pytorchbot merge

Manually re-landing as internal tweaks has been made
